monai/losses/dice.py

```diff
@@ -23,63 +23,82 @@
+        squared_pred=False,
+        jaccard=False
+    ):
         """
         Args:
             include_background (bool): If False channel index 0 (background category) is excluded from the calculation.
+            to_onehot_y (bool): whether to convert `y` into the one-hot format. Defaults to False.
             do_sigmoid (bool): If True, apply a sigmoid function to the prediction.
             do_softmax (bool): If True, apply a softmax function to the prediction.
+            squared_pred (bool): use squared versions of targets and predictions in the denominator or not.
+            jaccard (bool): compute Jaccard Index (soft IoU) instead of dice or not.
+
         """
         super().__init__()
         self.include_background = include_background
+        self.to_onehot_y = to_onehot_y
         if do_sigmoid and do_softmax:
             raise ValueError('do_sigmoid=True and do_softmax=Ture are not compatible.')
         self.do_sigmoid = do_sigmoid
         self.do_softmax = do_softmax
+        self.squared_pred = squared_pred
+        self.jaccard = jaccard
 
     def forward(self, pred, ground, smooth=1e-5):
-        if ground.shape[1] != 1:
-            raise ValueError("Ground truth should have only a single channel, shape is " + str(ground.shape))
-
-        psum = pred.float()
-        if self.do_sigmoid:
-            psum = psum.sigmoid()  # use sigmoid activation
-        if pred.shape[1] == 1:
-            if self.do_softmax:
-                raise ValueError('do_softmax is not compatible with single channel prediction.')
-            if not self.include_background:
+        """
+        Args:
+            pred (tensor): the shape should be BNH[WD].
+            ground (tensor): the shape should be BNH[WD].
+            smooth (float): a small constant to avoid nan.
+        """
+        n_pred_ch = pred.shape[1]
+        if self.do_sigmoid is True:
+            pred = torch.sigmoid(pred)
+        if self.do_softmax is True:
+            pred = torch.softmax(pred, 1)
+        if self.to_onehot_y is True:
+            if n_pred_ch == 1:
+                warnings.warn('single channel prediction, `to_onehot_y=True` ignored.')
+            else:
+                ground = one_hot(ground, n_pred_ch)
+
+        if self.include_background is False:
+            if n_pred_ch == 1:
                 warnings.warn('single channel prediction, `include_background=False` ignored.')
-            tsum = ground
-        else:  # multiclass dice loss
-            if self.do_softmax:
-                psum = torch.softmax(pred, 1)
-            tsum = one_hot(ground, pred.shape[1])  # B1HW(D) -> BNHW(D)
-            # exclude background category so that it doesn't overwhelm the other segmentations if they are small
-            if not self.include_background:
-                tsum = tsum[:, 1:]
-                psum = psum[:, 1:]
-        assert tsum.shape == psum.shape, ("Ground truth one-hot has differing shape (%r) from source (%r)" %
```

Comment:
i think this shape check is informative shouldn't be removed

Reply:
Added in the latest patch.
Thanks.